ci/lint.dart

```diff
@@ -0,0 +1,95 @@
+///
+/// usage:
+/// dart lint.dart <path to compile_commands.json> <path to git repository> [clang-tidy checks]
+
+import 'dart:io' show File, Process, ProcessResult, exit;
+import 'dart:convert' show jsonDecode, utf8;
+
+class Command {
+  String directory;
+  String command;
+  String file;
+}
+
+Command parseCommand(Map<String, dynamic> map) {
+  return Command()
+    ..directory = map['directory']
+    ..command = map['command']
+    ..file = map['file'];
+}
+
+String calcTidyArgs(Command command) {
+  String result = command.command;
+  result = result.replaceAll(RegExp(r'\S*clang/bin/clang'), '');
+  result = result.replaceAll(RegExp(r'-MF \S*'), '');
+  return result;
+}
+
+String calcTidyPath(Command command) {
+  final RegExp regex = RegExp(r'\S*clang/bin/clang');
+  return regex
+      .stringMatch(command.command)
+      .replaceAll('clang/bin/clang', 'clang/bin/clang-tidy');
+}
+
+bool isNonEmptyString(String str) => str.length > 0;
+
+bool containsAny(String str, List<String> queries) {
+  for (String query in queries) {
+    if (str.contains(query)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+/// Returns a list of all files with current changes or differ from `master`.
+List<String> getListOfChangedFiles(String repoPath) {
+  final Set<String> result = Set<String>();
+  final ProcessResult diffResult = Process.runSync(
+      'git', ['diff', '--name-only'],
+      workingDirectory: repoPath);
+  final ProcessResult diffCachedResult = Process.runSync(
+      'git', ['diff', '--cached', '--name-only'],
+      workingDirectory: repoPath);
+  final ProcessResult masterResult = Process.runSync(
+      'git', ['diff', '--name-only', 'master'],
+      workingDirectory: repoPath);
+  result.addAll(diffResult.stdout.split('\n').where(isNonEmptyString));
+  result.addAll(diffCachedResult.stdout.split('\n').where(isNonEmptyString));
+  result.addAll(masterResult.stdout.split('\n').where(isNonEmptyString));
+  return result.toList();
+}
+
+void main(List<String> arguments) async {
+  final String buildCommandsPath = arguments[0];
+  final String repoPath = arguments[1];
+  final String checks = arguments.length >= 3 ? arguments[2] : 'google-*';
+  final List<String> changedFiles = getListOfChangedFiles(repoPath);
+
+  final List<dynamic> buildCommandMaps =
+      jsonDecode(await new File(buildCommandsPath).readAsString());
+  final List<Command> buildCommands =
+      buildCommandMaps.map((x) => parseCommand(x)).toList();
+  final Command firstCommand = buildCommands[0];
+  final String tidyPath = calcTidyPath(firstCommand);
+  final List<Command> changedFileBuildCommands =
+      buildCommands.where((x) => containsAny(x.file, changedFiles)).toList();
+
+  int exitCode = 0;
+  for (Command command in changedFileBuildCommands) {
+    final String tidyArgs = calcTidyArgs(command);
+    print('# linting ${command.file}');
+    final List<String> args = [command.file, '--checks=$checks', '--'];
```

Comment:
You should be able to run a single clang-tidy invocation on all changed source files by specifying `-p out/compile_commands.json`

Reply:
yep looks possible, let me give it a shot.

Reply:
I spent a couple of minutes on this.  It is definitely possible.  It isn't a straight forward change since without parsing the compile_commands.json you have to decide which `clang-tidy` to use and the paths in the there don't match a known directory and you have to get them to match.  Suffice it to say it to say after running into a series of speed bumps it didn't seem worth the change now so I've left a TODO.

Reply:
```
YAML:1:1: error: not a mapping
-p /Users/aaclarke/dev/engine/src/out
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Error: invalid configuration specified.
Invalid argument

aaclarke-macbookpro2:src aaclarke$ ls /Users/aaclarke/dev/engine/src/out/compile_commands.json 
/Users/aaclarke/dev/engine/src/out/compile_commands.json
```
I tried one more time, for some reason clang-tidy doesn't like finding our compile_commands.json.